README.rst
=============
zabbix-client
=============

**zabbix-client** is a Zabbix API wrapper written in Python. It works on
Python 2.6+ and 3.2+.

Zabbix API
----------

Zabbix API was introduced in Zabbix 1.8 and allows you to create, update
and fetch Zabbix objects (like hosts, items, graphs and others) through
the JSON-RPC 2.0 protocol.

Zabbix API documentation:

* `Getting started with Zabbix API`_
* `Zabbix API description`_
* `Zabbix API method reference`_

JSON-RPC documentation:

* `JSON-RPC 2.0 specification`_

**zabbix-client** supports all Zabbix versions including the JSON-RPC
API, starting with Zabbix 1.8.

Usage
-----

Calling a method that does not require authentication::

    >>> from zabbix_client import ZabbixServerProxy
    >>> s = ZabbixServerProxy('http://localhost/zabbix')
    >>> s.apiinfo.version()
    '2.0.12'

Calling a method that requires previous authentication::

    >>> from zabbix_client import ZabbixServerProxy
    >>> s = ZabbixServerProxy('http://localhost/zabbix')
    >>> s.user.login(user='Admin', password='zabbix')
    '44cfb35933e3e75ef51988845ab15e8b'
    >>> s.host.get(output=['hostid', 'host'])
    [{'host': 'Zabbix server', 'hostid': '10084'},
        {'host': 'Test', 'hostid': '10085'}]
    >>> s.user.logout()
    True

License
-------

Licensed under the Apache License.

.. _Getting started with Zabbix API: https://www.zabbix.com/documentation/1.8/api/getting_started
.. _Zabbix API description: https://www.zabbix.com/documentation/2.2/manual/api
.. _Zabbix API method reference: https://www.zabbix.com/documentation/2.2/manual/api/reference
.. _JSON-RPC 2.0 specification: http://www.jsonrpc.org/specification


CHANGES.rst
Changes
=======

0.1.2 (2019-05-29)
------------------

* Add a missing required parameter on Zabbix 3 and later versions.

0.1.1 (2015-09-19)
------------------

* Fix exception handling.

0.1.0 (2014-08-10)
------------------

* Initial release.


zabbix_client/getter.py
# -*- coding: utf-8 -*-

import socket

from .exceptions import TransportError, TimeoutError
from .zabbix_protocol import NETWORK_TIMEOUT, ZabbixProtocolConnection


def dumps(key):
    return key.encode('utf-8') + b'\n'


def loads(response):
    return response.decode('utf-8').rstrip('\r\n')


class ZabbixGetter(object):

    def __init__(self, host, port=10050, timeout=NETWORK_TIMEOUT,
                 source_address=None):
        self.host = host
        self.port = port
        self.timeout = timeout
        if source_address is None:
            self.source_address = source_address
        else:
            self.source_address = (source_address, 0)

    def get(self, key):
        address = (self.host, self.port)
        connection = None
        try:
            connection = ZabbixProtocolConnection(
                address,
                timeout=self.timeout,
                source_address=self.source_address
            )

            try:
                connection.send(dumps(key), raw_value=True)
            except socket.timeout as e:
                raise TimeoutError(e)
            except socket.error as e:
                raise TransportError(e)

            return loads(connection.recv())
        finally:
            if connection:
                connection.close()


zabbix_client/sender.py
# -*- coding: utf-8 -*-

import socket
import json
import re
import logging

from .exceptions import (
    TransportError, TimeoutError, ResponseError, InvalidJSONError,
    ZabbixProtocolError
)
from .zabbix_protocol import NETWORK_TIMEOUT, ZabbixProtocolConnection


ZBX_RESPONSE_INFO_RE = (r'processed:? (\d+);? failed:? (\d+);? ' +
                        r'total:? (\d+);? seconds spent:? (\d+(?:\.\d{,6})?)')

logger = logging.getLogger(__name__)


def dumps(values):
    request = {
        'request': 'sender data',
        'data': values
    }
    data = json.dumps(request, separators=(',', ':')).encode('utf-8')

    if logger.isEnabledFor(logging.INFO):
        request_str = json.dumps(request, sort_keys=True)
        logger.info("Request: {0}".format(request_str))

    return data


def loads(response):
    try:
        zabbix_response = json.loads(response.decode('utf-8'))
    except ValueError as e:
        raise InvalidJSONError(e)

    if zabbix_response.get('response', '') != 'success':
        exception_message = 'Sender request failed: {0}'.format(zabbix_response)
        raise ZabbixProtocolError(exception_message, response=zabbix_response)

    if logger.isEnabledFor(logging.INFO):
        response_str = json.dumps(zabbix_response, sort_keys=True)
        logger.info("Response: {0}".format(response_str))

    info = zabbix_response.get('info', '')
    matched = re.match(ZBX_RESPONSE_INFO_RE, info, re.I)
    if not matched:
        raise ResponseError('Could not parse the response')

    return {
        'processed': int(matched.group(1)),
        'failed': int(matched.group(2)),
        'total': int(matched.group(3)),
        'seconds_spent': float(matched.group(4))
    }


class ZabbixSender(object):

    def __init__(self, server, port=10051, timeout=NETWORK_TIMEOUT,
                 source_address=None):
        self.server = server
        self.port = port
        self.timeout = timeout
        if source_address is None:
            self.source_address = source_address
        else:
            self.source_address = (source_address, 0)

    def send(self, values):
        if not values:
            return

        address = (self.server, self.port)
        connection = None
        try:
            connection = ZabbixProtocolConnection(
                address,
                timeout=self.timeout,
                source_address=self.source_address
            )

            try:
                connection.send(dumps(values))
            except socket.timeout as e:
                raise TimeoutError(e)
            except socket.error as e:
                raise TransportError(e)

            return loads(connection.recv())
        finally:
            if connection:
                connection.close()


zabbix_client/exceptions.py
# -*- coding: utf-8 -*-


class ZabbixClientError(Exception):
    pass


class TransportError(ZabbixClientError):
    pass


class TimeoutError(TransportError):
    pass


class HTTPError(TransportError):
    pass


class ResponseError(ZabbixClientError):
    pass


class ContentDecodingError(ResponseError):
    pass


class InvalidJSONError(ResponseError):
    pass


class JSONRPCError(ZabbixClientError):

    def __init__(self, *args, **kwargs):
        self.code = kwargs.pop('code', None)
        self.message = kwargs.pop('message', None)
        self.data = kwargs.pop('data', None)

        super(JSONRPCError, self).__init__(*args, **kwargs)


class ZabbixProtocolError(ZabbixClientError):

    def __init__(self, *args, **kwargs):
        self.response = kwargs.pop('response', None)

        super(ZabbixProtocolError, self).__init__(*args, **kwargs)


zabbix_client/api_wrapper.py
# -*- coding: utf-8 -*-

import logging
import json
import string
import socket

try:
    from urllib.request import Request, build_opener
except ImportError:
    # Python 2
    from urllib2 import Request, build_opener

try:
    from urllib.error import URLError, HTTPError as _HTTPError
except ImportError:
    # Python 2
    from urllib2 import URLError, HTTPError as _HTTPError

try:
    from io import BytesIO
except ImportError:
    # Python 2
    try:
        # C implementation
        from cStringIO import StringIO as BytesIO
    except ImportError:
        # Python implementation
        from StringIO import StringIO as BytesIO

try:
    import gzip
except ImportError:
    # Python can be built without zlib/gzip support
    gzip = None

try:
    import requests
except ImportError:
    requests = None

from . import __version__
from .exceptions import (
    ZabbixClientError, TransportError, TimeoutError, HTTPError, ResponseError,
    ContentDecodingError, InvalidJSONError, JSONRPCError
)


# Default network timeout (in seconds)
DEFAULT_TIMEOUT = 30

logger = logging.getLogger(__name__)


def dumps(id_, method, params=None, auth=None):
    rpc_request = {
        'jsonrpc': '2.0',
        'id': id_,
        'method': method
    }

    if params is not None:
        rpc_request['params'] = params
    else:
        # Zabbix 3 and later versions fail if 'params' is omitted
        rpc_request['params'] = {}
    if auth is not None:
        rpc_request['auth'] = auth

    dump = json.dumps(rpc_request, separators=(',', ':')).encode('utf-8')

    if logger.isEnabledFor(logging.INFO):
        json_str = json.dumps(rpc_request, sort_keys=True)
        logger.info("JSON-RPC request: {0}".format(json_str))

    return dump


def loads(response):
    try:
        rpc_response = json.loads(response.decode('utf-8'))
    except ValueError as e:
        raise InvalidJSONError(e)

    if not isinstance(rpc_response, dict):
        raise ResponseError('Response is not a dict')

    if 'jsonrpc' not in rpc_response or rpc_response['jsonrpc'] != '2.0':
        raise ResponseError('JSON-RPC version not supported')

    if 'error' in rpc_response:
        error = rpc_response['error']
        if 'code' not in error or 'message' not in error:
            raise ResponseError('Invalid JSON-RPC error object')

        code = error['code']
        message = error['message']
        # 'data' may be omitted
        data = error.get('data', None)

        if data is None:
            exception_message = 'Code: {0}, Message: {1}'.format(code, message)
        else:
            exception_message = ('Code: {0}, Message: {1}, ' +
                                 'Data: {2}').format(code, message, data)
        raise JSONRPCError(exception_message, code=code, message=message,
                           data=data)

    if 'result' not in rpc_response:
        raise ResponseError('Response does not contain a result object')

    if logger.isEnabledFor(logging.INFO):
        json_str = json.dumps(rpc_response, sort_keys=True)
        logger.info("JSON-RPC response: {0}".format(json_str))

    return rpc_response


class ZabbixServerProxy(object):

    def __init__(self, url, transport=None):
        self.url = url if not url.endswith('/') else url[:-1]
        self.url += '/api_jsonrpc.php'

        logger.debug("Zabbix server URL: {0}".format(self.url))

        if transport is not None:
            self.transport = transport
        else:
            if requests:
                logger.debug("Using requests as transport layer")
                self.transport = RequestsTransport()
            else:
                logger.debug("Using urllib libraries as transport layer")
                self.transport = UrllibTransport()

        self._request_id = 0
        self._auth_token = None

        self._method_hooks = {
            'apiinfo.version': self._no_auth_method,
            'user.login': self._login,
            'user.authenticate': self._login, # deprecated alias of user.login
            'user.logout': self._logout
        }

    def __getattr__(self, name):
        return ZabbixObject(name, self)

    def call(self, method, params=None):
        method_lower = method.lower()

        if method_lower in self._method_hooks:
            return self._method_hooks[method_lower](method, params=params)

        return self._call(method, params=params, auth=self._auth_token)

    def _call(self, method, params=None, auth=None):
        self._request_id += 1
        rpc_request = dumps(self._request_id, method, params=params, auth=auth)

        content = self.transport.request(self.url, rpc_request)

        rpc_response = loads(content)

        return rpc_response['result']

    def _no_auth_method(self, method, params=None):
        return self._call(method, params=params)

    def _login(self, method, params=None):
        self._auth_token = None

        # Save the new token if the request is successful
        self._auth_token = self._call(method, params=params)

        return self._auth_token

    def _logout(self, method, params=None):
        try:
            result = self._call(method, params=params, auth=self._auth_token)
        except ZabbixClientError:
            raise
        finally:
            self._auth_token = None

        return result


class ZabbixObject(object):

    def __init__(self, name, server_proxy):
        self.name = name
        self.server_proxy = server_proxy

    def __getattr__(self, name):
        def call_wrapper(*args, **kwargs):
            if args and kwargs:
                raise ValueError('JSON-RPC 2.0 does not allow both ' +
                                 'positional and keyword arguments')

            method = '{0}.{1}'.format(self.name, name)
            params = args or kwargs or None
            return self.server_proxy.call(method, params=params)

        # Little hack to avoid clashes with reserved keywords.
        # Example: use configuration.import_() to call configuration.import()
        if name.endswith('_'):
            name = name[:-1]

        return call_wrapper


class Transport(object):

    def __init__(self, timeout=DEFAULT_TIMEOUT):
        self.timeout = timeout

    def request(self, url, rpc_request):
        raise NotImplementedError

    @staticmethod
    def _add_headers(headers):
        # Set the JSON-RPC headers
        json_rpc_headers = {
            'Content-Type': 'application/json',
            'Accept': 'application/json'
        }
        headers.update(json_rpc_headers)

        # If no custom header exists, set the default user-agent
        if 'User-Agent' not in headers:
            headers['User-Agent'] = 'zabbix-client/{0}'.format(__version__)


class RequestsTransport(Transport):

    def __init__(self, *args, **kwargs):
        if not requests:
            raise ValueError('requests is not available')

        self.session = kwargs.pop('session', None)

        super(RequestsTransport, self).__init__(*args, **kwargs)

        if self.session is None:
            self.session = requests.Session()

            # Delete default requests' user-agent
            self.session.headers.pop('User-Agent', None)

        self._add_headers(self.session.headers)

    def request(self, url, rpc_request):
        try:
            response = self.session.post(url, data=rpc_request,
                                         timeout=self.timeout)

            response.raise_for_status()

            content = response.content
        except requests.Timeout as e:
            raise TimeoutError(e)
        except requests.exceptions.ContentDecodingError as e:
            raise ContentDecodingError(e)
        except requests.HTTPError as e:
            raise HTTPError(e)
        except requests.RequestException as e:
            raise TransportError(e)

        return content


class UrllibTransport(Transport):

    def __init__(self, *args, **kwargs):
        self.accept_gzip_encoding = kwargs.pop('accept_gzip_encoding', True)
        headers = kwargs.pop('headers', None)

        super(UrllibTransport, self).__init__(*args, **kwargs)

        self.headers = {}
        if headers:
            for key, value in headers.items():
                self.headers[string.capwords(key, '-')] = value

        self._add_headers(self.headers)

        if self.accept_gzip_encoding and gzip:
            self.headers['Accept-Encoding'] = 'gzip'

        self._opener = build_opener()

    def request(self, url, rpc_request):
        request = Request(url, data=rpc_request, headers=self.headers)

        try:
            response = self._opener.open(request, timeout=self.timeout)

            content = response.read()
        except _HTTPError as e:
            raise HTTPError(e)
        except URLError as e:
            if isinstance(e.reason, socket.timeout):
                raise TimeoutError(e)
            else:
                raise TransportError(e)
        except socket.timeout as e:
            raise TimeoutError(e)

        encoding = response.info().get('Content-Encoding', '').lower()
        if encoding in ('gzip', 'x-gzip'):
            if not gzip:
                raise ValueError('gzip is not available')

            b = BytesIO(content)
            try:
                content = gzip.GzipFile(mode='rb', fileobj=b).read()
            except IOError as e:
                raise ContentDecodingError(e)

        return content


zabbix_client/zabbix_protocol.py
# -*- coding: utf-8 -*-

import socket
import struct
import sys

from .exceptions import TransportError, TimeoutError, ResponseError
from .utils import recv_bytes


# Default network timeout (in seconds)
NETWORK_TIMEOUT = 30

# Maximum response size: 128 MiB
ZBX_MAX_RECV_DATA_SIZE = 128 * 1024 * 1024


class ZabbixProtocol(object):

    ZBX_TCP_HEADER_DATA = b'ZBXD'
    ZBX_TCP_HEADER_VERSION = 1
    ZBX_TCP_HEADER = struct.pack('<4sB', ZBX_TCP_HEADER_DATA,
                                 ZBX_TCP_HEADER_VERSION)

    @staticmethod
    def create_connection(address, timeout=None, source_address=None):
        try:
            if sys.version_info < (2, 7):
                return socket.create_connection(address, timeout=timeout)
            else:
                return socket.create_connection(address, timeout=timeout,
                                                source_address=source_address)
        except socket.timeout as e:
            raise TimeoutError(e)
        except (socket.error, socket.gaierror) as e:
            raise TransportError(e)

    @staticmethod
    def close_connection(connection):
        connection.close()

    @classmethod
    def send(cls, sock, data, raw_value=False):
        if raw_value:
            zabbix_request = data
        else:
            data_len = struct.pack('<Q', len(data))
            zabbix_request = cls.ZBX_TCP_HEADER + data_len + data

        try:
            sock.sendall(zabbix_request)
        except socket.timeout as e:
            raise TimeoutError(e)
        except socket.error as e:
            raise TransportError(e)

    @classmethod
    def recv(cls, sock):
        try:
            header = recv_bytes(sock, len(cls.ZBX_TCP_HEADER))
            if not header:
                raise ResponseError('Empty response')
            if header != cls.ZBX_TCP_HEADER:
                raise ResponseError('Wrong header')

            data_len_chunk = recv_bytes(sock, 8)
            if len(data_len_chunk) < 8:
                raise ResponseError('Response is shorter than expected')

            data_len = struct.unpack('<Q', data_len_chunk)[0]
            if ZBX_MAX_RECV_DATA_SIZE < data_len:
                raise ResponseError('Response exceeds the maximum allowed size')

            data = sock.recv(data_len)
            if len(data) < data_len:
                raise ResponseError('Response is shorter than expected')

            return data
        except socket.timeout as e:
            raise TimeoutError(e)
        except socket.error as e:
            raise TransportError(e)


class ZabbixProtocolConnection(object):

    def __init__(self, address, timeout=None, source_address=None):
        self.address = address
        self.timeout = timeout
        self.source_address = source_address

        self._socket = ZabbixProtocol.create_connection(
            self.address,
            timeout=self.timeout,
            source_address=self.source_address
        )

    def send(self, data, raw_value=False):
        ZabbixProtocol.send(self._socket, data, raw_value=raw_value)

    def recv(self):
        return ZabbixProtocol.recv(self._socket)

    def close(self):
        ZabbixProtocol.close_connection(self._socket)


zabbix_client/__init__.py
# -*- coding: utf-8 -*-

"""
zabbix-client
=============

zabbix-client is a Zabbix API wrapper written in Python. Zabbix API was
introduced in Zabbix 1.8 and allows you to create, update and fetch
Zabbix objects (like hosts, items, graphs and others) through the
JSON-RPC 2.0 protocol.

Zabbix API documentation:

* https://www.zabbix.com/documentation/1.8/api/getting_started
* https://www.zabbix.com/documentation/2.2/manual/api
* https://www.zabbix.com/documentation/2.2/manual/api/reference

JSON-RPC 2.0 specification:

* http://www.jsonrpc.org/specification

Usage
-----

Calling a method that does not require authentication::

    >>> from zabbix_client import ZabbixServerProxy
    >>> s = ZabbixServerProxy('http://localhost/zabbix')
    >>> s.apiinfo.version()
    '2.0.12'

Calling a method that requires previous authentication::

    >>> from zabbix_client import ZabbixServerProxy
    >>> s = ZabbixServerProxy('http://localhost/zabbix')
    >>> s.user.login(user='Admin', password='zabbix')
    '44cfb35933e3e75ef51988845ab15e8b'
    >>> s.host.get(output=['hostid', 'host'])
    [{'host': 'Zabbix server', 'hostid': '10084'},
        {'host': 'Test', 'hostid': '10085'}]
    >>> s.user.logout()
    True

"""

__version__ = '0.1.2'
__author__ = 'Jesús Losada Novo'
__license__ = 'Apache 2.0'
__copyright__ = 'Copyright 2014-2015 Jesús Losada Novo'


from .api_wrapper import (
    ZabbixServerProxy, Transport, RequestsTransport, UrllibTransport
)
from .exceptions import (
    ZabbixClientError, TransportError, TimeoutError, HTTPError, ResponseError,
    ContentDecodingError, InvalidJSONError, JSONRPCError
)


# Add a do-nothing logging handler.
# https://docs.python.org/3/howto/logging.html#library-config
import logging
try:
    from logging import NullHandler
except ImportError:
    class NullHandler(logging.Handler):
        def emit(self, record):
            pass

logging.getLogger(__name__).addHandler(NullHandler())
